client/maisie/resources/models.py

```diff
@@ -0,0 +1,132 @@
+        dataset_name: str = "",
+        dataset_description: str = "",
+    ):
+        hyperparameters = self._determine_input(hyperparameters)
+        parameters = self._determine_input(parameters)
+        metrics = self._determine_input(metrics)
+
+        with self.config.session as session:
+            files = {}
+            try:
+                files["file"] = open(filename, "rb")
+            except FileNotFoundError:
+                logger.error(f"Model `{filename}` could not be found.")
+
+            git = GitProvider(self.config.git_local_repo)
+            payload = {
+                "name": name,
+                "hyperparameters": json.dumps(hyperparameters),
+                "parameters": json.dumps(parameters),
+                "metrics": json.dumps(metrics),
+                "private": private,
+                "user_id": 1,
+                "project_id": self.config.selected_project,
+                "git_active_branch": git.active_branch,
+                "git_commit_hash": git.latest_commit,
+            }
+            request = session.post(
+                f"{self.config.api_url}/models/", files=files, data=payload
+            )
+
+            results = []
+            # print(payload)
+            # print(request.text)
+            if "data" in request.json():
+                results.append(request.json()["data"])
+            else:
+                logger.error("Could not upload selected model.")
+
+            return results
+
+    def update(self, id: int, data: dict):
+        with self.config.session as session:
+            pass
+
+    def download(self, id: int):
+        with self.config.session as session:
+            request = session.get(f"{self.config.api_url}/models/{id}/")
+            if ("data") in request.json():
+                data = request.json()["data"]
+                if "_links" in data:
+                    links = data["_links"]
+                    if "download" in links:
+                        download_link = links["download"]
+                        opener = urllib.request.build_opener()
+                        opener.addheaders = [
+                            (
+                                "User-Agent",
+                                "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/36.0.1941.0 Safari/537.36",
+                            )
+                        ]
+                        urllib.request.install_opener(opener)
+                        download_data = urllib.request.urlretrieve(
+                            download_link, f"model_{id}"
+                        )
+                        # The return value is a tuple consisting of a local filename and either
+                        # a mimetools.Message object containing the response headers (for remote URLs) or None (for local URLs).
+                        download_data = session.get(download_link)
+                        return download_data
```

Comment:
If I read the implementation correctly, this method does not save the requested models locally, which is the whole point of this procedure. 
`Models.download()` can take some assumptions about the naming scheme or ask the user for input (implementing both solutions would be ideal).
This method is supposed to use the provided `download_link` to fetch the resource an save it to our local drive, not display it's body or the response headers and metadata.